src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -176,52 +243,162 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
+            static ImmutableArray<string> AttachComplexTypes(ImmutableArray<string> receiverTypeNames)
+            {
+                using var _ = ArrayBuilder<string>.GetInstance(receiverTypeNames.Length + 2, out var receiverTypeNamesBuilder);
+                receiverTypeNamesBuilder.AddRange(receiverTypeNames);
+                receiverTypeNamesBuilder.Add(FindSymbols.Extensions.ComplexReceiverTypeName);
+                receiverTypeNamesBuilder.Add(FindSymbols.Extensions.ComplexArrayReceiverTypeName);
+
+                return receiverTypeNamesBuilder.ToImmutable();
+            }
+        }
+
+        private static void GetExtensionMethodItemsForSymbolsFromDifferentCompilation(
+            int position,
+            SemanticModel semanticModel,
+            ITypeSymbol receiverTypeSymbol,
+            MultiDictionary<ITypeSymbol, IMethodSymbol> matchingMethodSymbols,
+            ArrayBuilder<SerializableImportCompletionItem> builder,
+            Dictionary<INamespaceSymbol, string> stringCache,
+            Dictionary<ITypeSymbol, bool> checkedReceiverTypes,
+            CancellationToken cancellationToken)
+        {
+            // Matching extension method symbols are grouped based on their receiver type.
+            foreach (var (declaredReceiverType, methodSymbols) in matchingMethodSymbols)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var declaredReceiverTypeInCurrentCompilation = SymbolFinder.FindSimilarSymbols(declaredReceiverType, semanticModel.Compilation).FirstOrDefault();
+                if (declaredReceiverTypeInCurrentCompilation == null)
+                {
+                    // Bug: https://github.com/dotnet/roslyn/issues/45404
+                    // SymbolFinder.FindSimilarSymbols would fail if current and referenced compilation targeting different frameworks say net472 and netstandard respectively.
+                    // Here's SymbolKey for System.String from those two framework as an example:
+                    //
+                    //  {1 (D "String" (N "System" 0 (N "" 0 (U (S "netstandard" 4) 3) 2) 1) 0 0 (% 0) 0)}
+                    //  {1 (D "String"(N "System" 0(N "" 0(U(S "mscorlib" 4) 3) 2) 1) 0 0(% 0) 0)}
+                    //
+                    // Also we don't use the "ignoreAssemblyKey" option for SymbolKey resolution because its perfermance doesn't meet our requirement.
+                    continue;
+                }
+
+                if (checkedReceiverTypes.TryGetValue(declaredReceiverTypeInCurrentCompilation, out var cachedResult) && !cachedResult)
+                {
+                    // If we already checked an extension method with same receiver type before, and we know it can't be applied
+                    // to the receiverTypeSymbol, then no need to proceed methods from this group..
+                    continue;
+                }
+
```

Comment:
i would also recommend extracing from here to the end of the loop into a new helper for clarity.

Reply:
I tried, but the long parameter list feels bulky:

```cs
            static void ProcessMethodSymbols(
                int position,
                SemanticModel semanticModel,
                ITypeSymbol receiverTypeSymbol,
                ITypeSymbol declaredReceiverTypeInCurrentCompilation, bool isReceiverTypeKnownMatch,
                MultiDictionary<ITypeSymbol, IMethodSymbol>.ValueSet methodSymbols,
                ArrayBuilder<SerializableImportCompletionItem> builder,
                Dictionary<INamespaceSymbol, string> stringCache,
                Dictionary<ITypeSymbol, bool> checkedReceiverTypes)
            {
                // This is also affected by the symbol resolving issue mentioned above, which means in case referenced projects
                // are targeting different framework, we will miss extension methods with any framework type in their signature from those projects.
                var isFirstMethod = true;
                foreach (var methodInCurrentCompilation in methodSymbols.Select(s => SymbolFinder.FindSimilarSymbols(s, semanticModel.Compilation).FirstOrDefault()).WhereNotNull())
                {
                    if (isFirstMethod)
                    {
                        isFirstMethod = false;

                        // We haven't seen this receiver type yet. Try to check by reducing one extension method
                        // to the given receiver type and save the result.
                        if (!isReceiverTypeKnownMatch)
                        {
                            // If this is the first symbol we retrived from current compilation,
                            // try to check if we can apply it to given receiver type, and save result to our cache.
                            var reducedMethodSymbol = methodInCurrentCompilation.ReduceExtensionMethod(receiverTypeSymbol);
                            var isReceiverTypeMatch = reducedMethodSymbol != null;
                            checkedReceiverTypes[declaredReceiverTypeInCurrentCompilation] = isReceiverTypeMatch;

                            // Now, isReceiverTypeMatch being false means method doesn't match the receiver type,
                            // stop processing methods from this group.
                            if (!isReceiverTypeMatch)
                            {
                                return;
                            }
                        }
                    }

                    if (semanticModel.IsAccessible(position, methodInCurrentCompilation))
                    {
                        CreateAndAddItem(methodInCurrentCompilation, builder, stringCache);
                    }
                }
```

Reply:
yup.  that does seem bulky.  retracted.